src/multi/mod.rs

```diff
@@ -55,36 +59,61 @@ const MAX_INITIAL_CAPACITY_BYTES: usize = 65536;
 #[cfg(feature = "alloc")]
 #[cfg_attr(feature = "docsrs", doc(cfg(feature = "alloc")))]
 pub fn many0<I, F>(
-  mut f: F,
-) -> impl FnMut(I) -> IResult<I, Vec<<F as Parser<I>>::Output>, <F as Parser<I>>::Error>
+  f: F,
+) -> impl Parser<I, Output = Vec<<F as Parser<I>>::Output>, Error = <F as Parser<I>>::Error>
 where
   I: Clone + InputLength,
   F: Parser<I>,
 {
-  move |mut i: I| {
-    let mut acc = crate::lib::std::vec::Vec::with_capacity(4);
+  Many0 { parser: f }
+}
+
+/// Prser implementation for the [many0] combinator
+pub struct Many0<F> {
+  parser: F,
+}
```

Comment:
In winnow-rs/winnow#6, I gave up on moving the parsers to be structs in part because of type inference issues.  Looking back at the errors, it looks like it all came down to my `PhantomData`:
```
error[E0282]: type annotations needed
Error:     --> src/number/complete.rs:1411:9
     |
1411 |         map(tuple((digit1, opt(pair(char('.'), opt(digit1))))), |_| ()),
     |         ^^^ cannot infer type of the type parameter `PO` declared on the function `map`
     |
help: consider specifying the generic arguments
     |
1411 |         map::<_, PO, _, I, O, E>(tuple((digit1, opt(pair(char('.'), opt(digit1))))), |_| ()),
     |            +++++++++++++++++++++
```
*(`PO`, `I`, `O`, and `E` were `PhantomData` fields)*

I suspect rust-bakery/nom#1626 would help with this type inference problem and, now looking at it, I think it does, making life easier on this PR than it was for me on winnow-rs/winnow#6.